internal/tests/tests/ue/downlink_data_paging.go

```diff
@@ -0,0 +1,363 @@
+		return fmt.Errorf("initial registration procedure failed: %v", err)
+	}
+
+	logger.Logger.Debug(
+		"Completed Initial Registration Procedure",
+		zap.String("IMSI", newUE.UeSecurity.Supi),
+		zap.Int64("RAN UE NGAP ID", RANUENGAPID),
+		zap.Int64("AMF UE NGAP ID", gNodeB.GetAMFUENGAPID(RANUENGAPID)),
+	)
+
+	ueIP := newUE.GetPDUSession().UEIP + "/16"
+
+	gnbPDUSession := gNodeB.GetPDUSession(RANUENGAPID)
+
+	tunInterfaceName := GTPInterfaceNamePrefix + "paging0"
+
+	_, err = gNodeB.AddTunnel(&gnb.NewTunnelOpts{
+		UEIP:             ueIP,
+		UpfIP:            gnbPDUSession.UpfAddress,
+		TunInterfaceName: tunInterfaceName,
+		ULteid:           gnbPDUSession.ULTeid,
+		DLteid:           gnbPDUSession.DLTeid,
+	})
+	if err != nil {
+		return fmt.Errorf("could not create GTP tunnel (name: %s, DL TEID: %d): %v", tunInterfaceName, gnbPDUSession.DLTeid, err)
+	}
+
+	logger.GnbLogger.Debug(
+		"Created GTP Tunnel for PDU Session",
+		zap.String("Interface", tunInterfaceName),
+		zap.String("UE IP", ueIP),
+		zap.String("UPF IP", gnbPDUSession.UpfAddress),
+		zap.Uint32("UL TEID", gnbPDUSession.ULTeid),
+		zap.Uint32("DL TEID", gnbPDUSession.DLTeid),
+	)
+
+	cmd := exec.Command("ping", "-I", tunInterfaceName, env.Config.Subscriber.PingDestination, "-c", "3", "-W", "1")
+
+	err = cmd.Run()
+	if err != nil {
+		return fmt.Errorf("could not ping destination %s: %v", env.Config.Subscriber.PingDestination, err)
+	}
+
+	logger.Logger.Debug(
+		"Ping successful",
+		zap.String("interface", tunInterfaceName),
+		zap.String("destination", env.Config.Subscriber.PingDestination),
+	)
+
+	pduSessionStatus := [16]bool{}
+	pduSessionStatus[PDUSessionID] = true
+
+	err = procedure.UEContextRelease(&procedure.UEContextReleaseOpts{
+		AMFUENGAPID:   gNodeB.GetAMFUENGAPID(RANUENGAPID),
+		RANUENGAPID:   RANUENGAPID,
+		GnodeB:        gNodeB,
+		UE:            newUE,
+		PDUSessionIDs: pduSessionStatus,
+	})
+	if err != nil {
+		return fmt.Errorf("UEContextReleaseProcedure failed: %v", err)
+	}
+
+	err = gNodeB.CloseTunnel(gnbPDUSession.DLTeid)
+	if err != nil {
+		return fmt.Errorf("could not close GTP tunnel (name: %s, DL TEID: %d): %v", tunInterfaceName, gnbPDUSession.DLTeid, err)
+	}
+
+	source_ip, err := net.ResolveIPAddr("ip", env.Config.Gnb.N3Address)
+	if err != nil {
+		return fmt.Errorf("could not parse N3 IP address: %v", err)
+	}
+
+	n3_interface, err := scago_utils.GetInterfaceByIP(source_ip.IP)
+	if err != nil {
+		return fmt.Errorf("could not find N3 interface from IP address: %v", err)
+	}
+
+	n3_mac := scago_utils.MacByInt(n3_interface.Name)
+	eth := packet.EthernetLayer()
+
+	if err = eth.SetSrcMAC(n3_mac); err != nil {
+		return fmt.Errorf("could not set source MAC: %v", err)
+	}
+
+	if err = eth.SetDstMAC("a8:b8:e0:03:df:5c"); err != nil {
+		return fmt.Errorf("could not set destination MAC: %v", err)
+	}
+
+	eth.SetEthernetType(layers.EthernetTypeIPv4)
+
+	ip := packet.IPv4Layer()
+
+	if err = ip.SetDstIP(newUE.GetPDUSession().UEIP); err != nil {
```

Comment:
I don't quite understand how this packet makes its way to Ella Core, is that because we set the mac address to Ella Core's mac?